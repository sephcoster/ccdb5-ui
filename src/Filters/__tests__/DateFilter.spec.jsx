import ReduxDateFilter, { DateFilter, mapDispatchToProps } from '../DateFilter'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme';
import thunk from 'redux-thunk'

function setupEnzyme(initialProps={}) {
  const props = Object.assign({
    changeDateRange: jest.fn(),
    fieldName: 'date_received',
    from: '',
    maximumDate: null,
    minimumDate: null,
    through: '',
    title: 'Date CFPB Received the complaint'
  }, initialProps)

  const target = mount(<DateFilter {...props} />);

  return {
    props,
    target
  }
}

function setupSnapshot(query={}) {
  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore({
    query
  })

  return renderer.create(
    <Provider store={store}>
      <ReduxDateFilter fieldName="date_received"
                       minimumDate={ new Date( '2015-01-01T12:00:00.000Z' ) }
                       maximumDate={ new Date( '2015-12-31T12:00:00.000Z' ) }
                       title="Date CFPB Received the complaint"
      />
    </Provider>
  )
}

describe('component::DateFilter', () => {
  describe('snapshot', () => {
    it('supports no dates in the Redux Store', () => {
      const target = setupSnapshot()
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('shows errors', () => {
      const target = setupSnapshot({
        date_received_min: new Date(2016, 0, 1),
        date_received_max: new Date(2000, 0, 1)
      })
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('DateInput interface', () => {
    let target, props
    beforeEach(() => {
      ({ target, props } = setupEnzyme())
    })

    describe('_onDateEntered', () => {
      it('converts the date to a string', () => {
        target.instance()._onDateEntered('from', new Date(2012, 11, 31))
        expect(target.state('from')).toEqual('12/31/2012')
      })

      it('resets any messages that may have existed', () => {
        target.setState({
          messages: {
            from: 'foo'
          }
        })
        target.instance()._onDateEntered('from', new Date(2012, 11, 31))
        expect(target.state('messages')).toEqual({})
      })

      it('checks for range errors', () => {
        target.setState({
          through: '1/1/2000'
        })
        target.instance()._onDateEntered('from', new Date(2012, 11, 31))
        expect(target.state('messages').ordered).toEqual(
          '\'From\' must be less than \'Through\''
        )
      })

      it('calls changeDateRange', () => {
        target.instance()._onDateEntered('through', new Date(2012, 11, 31))
        expect(props.changeDateRange).toHaveBeenCalledWith(
          null, new Date(2012, 11, 31)
        )
      })
    })

    describe('_onError', () => {
      it('associates the error to the field', () => {
        target.instance()._onError('from', 'foo', '2/31/2012')
        expect(target.state()).toEqual({
          from: '2/31/2012',
          messages: {
            from: 'foo'
          },
          through: ''
        })
      })
    })
  })

  describe('componentWillReceiveProps', () => {
    it('does not trigger a new update', () => {
      const {target, props} = setupEnzyme()
      target.setProps({from: '2016-01-01', through: '2015-12-31'})
      expect(props.changeDateRange).not.toHaveBeenCalled()
    })    
  })

  describe('mapDispatchToProps', () => {
    it('hooks into changeDateRange', () => {
      const dispatch = jest.fn()
      const props = {fieldName: 'qaz'}
      mapDispatchToProps(dispatch, props).changeDateRange('foo', 'bar')
      expect(dispatch.mock.calls.length).toEqual(1)
    })
  })
})
