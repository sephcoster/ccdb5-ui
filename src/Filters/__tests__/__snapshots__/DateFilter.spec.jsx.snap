// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::DateFilter snapshot shows errors 1`] = `
<section
  className="o-expandable aggregation date-filter"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      Date CFPB Received the complaint
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <span
          className="cf-icon cf-icon-minus-round"
        />
      </button>
    </span>
  </div>
  <p />
  <div>
    <ul
      className="date-inputs"
    >
      <li>
        <label
          className="a-label a-label__heading body-copy"
        >
          From
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            aria-describedby="input-error_message-from"
            className="a-text-input"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="mm/dd/yyyy"
            size="15"
            value="1/1/2016"
          />
          <button
            className="a-btn a-btn__link"
            onClick={[Function]}
          >
            <span
              className="cf-icon cf-icon-delete"
            />
          </button>
        </div>
      </li>
      <li>
        <label
          className="a-label a-label__heading body-copy"
        >
          Through
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            aria-describedby="input-error_message-through"
            className="a-text-input"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="mm/dd/yyyy"
            size="15"
            value="1/1/2000"
          />
          <button
            className="a-btn a-btn__link"
            onClick={[Function]}
          >
            <span
              className="cf-icon cf-icon-delete"
            />
          </button>
        </div>
      </li>
    </ul>
    <ul
      className="messages"
    >
      <li
        className="a-error-message"
        id="input-error_message-ordered"
        role="alert"
      >
        <span
          aria-hidden="true"
          className="cf-icon cf-icon-delete-round"
        />
        'From' must be less than 'Through'
      </li>
      <li
        className="a-error-message"
        id="input-error_message-through"
        role="alert"
      >
        <span
          aria-hidden="true"
          className="cf-icon cf-icon-delete-round"
        />
        '1/1/2000' must be greater than 1/1/2015
      </li>
    </ul>
  </div>
</section>
`;

exports[`component::DateFilter snapshot supports no dates in the Redux Store 1`] = `
<section
  className="o-expandable aggregation date-filter"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      Date CFPB Received the complaint
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <span
          className="cf-icon cf-icon-minus-round"
        />
      </button>
    </span>
  </div>
  <p />
  <div>
    <ul
      className="date-inputs"
    >
      <li>
        <label
          className="a-label a-label__heading body-copy"
        >
          From
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            aria-describedby="input-error_message-from"
            className="a-text-input"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="mm/dd/yyyy"
            size="15"
            value=""
          />
        </div>
      </li>
      <li>
        <label
          className="a-label a-label__heading body-copy"
        >
          Through
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            aria-describedby="input-error_message-through"
            className="a-text-input"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="mm/dd/yyyy"
            size="15"
            value=""
          />
        </div>
      </li>
    </ul>
  </div>
</section>
`;
